File: python-backend/app/utils/work_schedule.py
from datetime import datetime


DEFAULT_WORK_SCHEDULE = {
    "start_time": "09:00",
    "end_time": "18:00",
    "timezone": "Asia/Calcutta"
}


def parse_time_to_minutes(value: str | None) -> int:
    if not value or ":" not in value:
        return 0
    hours, minutes = value.split(":")
    try:
        return int(hours) * 60 + int(minutes)
    except ValueError:
        return 0


def get_expected_work_hours(work_schedule: dict | None) -> float:
    schedule = work_schedule or DEFAULT_WORK_SCHEDULE
    duration = max(0, parse_time_to_minutes(schedule.get("end_time")) - parse_time_to_minutes(schedule.get("start_time")))
    return duration / 60


def get_worked_hours(check_in: datetime | None, check_out: datetime | None) -> float:
    if not check_in or not check_out:
        return 0
    return max(0, (check_out - check_in).total_seconds() / 3600)


def build_session_summary(records: list[dict]) -> dict:
    sorted_records = sorted(records, key=lambda item: item["timestamp"])
    sessions = []
    open_check_in = None

    for record in sorted_records:
        if record["type"] == "check_in":
            open_check_in = record["timestamp"]
            continue
        if record["type"] == "check_out" and open_check_in:
            sessions.append({
                "check_in": open_check_in,
                "check_out": record["timestamp"],
                "worked_hours": get_worked_hours(open_check_in, record["timestamp"])
            })
            open_check_in = None

    first_session = sessions[0] if sessions else None
    last_session = sessions[-1] if sessions else None
    worked_hours = sum(session["worked_hours"] for session in sessions)

    return {
        "sessions": sessions,
        "session_count": len(sessions),
        "check_in": first_session["check_in"] if first_session else open_check_in,
        "check_out": last_session["check_out"] if last_session else None,
        "worked_hours": worked_hours,
        "in_progress": bool(open_check_in)
    }


def build_daily_work_metrics(records: list[dict], work_schedule: dict | None) -> dict:
    expected_hours = get_expected_work_hours(work_schedule)
    session_summary = build_session_summary(records)
    overtime_hours = max(0, session_summary["worked_hours"] - expected_hours)
    return {
        **session_summary,
        "expected_hours": round(expected_hours, 1),
        "worked_hours": round(session_summary["worked_hours"], 1),
        "overtime_hours": round(overtime_hours, 1)
    }

File: node-backend/src/utils/workSchedule.js
export const DEFAULT_WORK_SCHEDULE = {
  start_time: "09:00",
  end_time: "18:00",
  timezone: "Asia/Calcutta"
};

const parseMinutes = (value = "") => {
  const [hours, minutes] = String(value).split(":").map(Number);
  if (!Number.isFinite(hours) || !Number.isFinite(minutes)) {
    return 0;
  }
  return hours * 60 + minutes;
};

export const expectedHours = (schedule = DEFAULT_WORK_SCHEDULE) =>
  Math.max(0, parseMinutes(schedule.end_time) - parseMinutes(schedule.start_time)) / 60;

export const buildDailyMetrics = (records, schedule = DEFAULT_WORK_SCHEDULE) => {
  const sorted = [...records].sort((a, b) => a.timestamp - b.timestamp);
  const sessions = [];
  let openCheckIn = null;

  for (const record of sorted) {
    if (record.type === "check_in") {
      openCheckIn = record.timestamp;
    } else if (record.type === "check_out" && openCheckIn) {
      const worked = Math.max(0, (record.timestamp - openCheckIn) / 3600000);
      sessions.push({ check_in: openCheckIn, check_out: record.timestamp, worked_hours: worked });
      openCheckIn = null;
    }
  }

  const workedHours = sessions.reduce((total, session) => total + session.worked_hours, 0);
  const expected = expectedHours(schedule);
  return {
    sessions,
    session_count: sessions.length,
    check_in: sessions[0]?.check_in || openCheckIn,
    check_out: sessions[sessions.length - 1]?.check_out || null,
    worked_hours: Math.round(workedHours * 10) / 10,
    expected_hours: Math.round(expected * 10) / 10,
    overtime_hours: Math.round(Math.max(0, workedHours - expected) * 10) / 10,
    in_progress: Boolean(openCheckIn)
  };
};
